File: java/permutations.java
class Solution {
    
    int[] nums = new int[0];
    List<List<Integer>> solutions = new ArrayList();
    Set<Integer> unusedIndices = new HashSet();
    Stack<Integer> stack = new Stack();
    
    public List<List<Integer>> permute(int[] nums) {
        this.nums = nums;
        for (int i = 0; i < nums.length; i++) {
            unusedIndices.add(i);
        }
        backtrack();
        return solutions;
    }
    
    private void backtrack() {
        if (unusedIndices.size() == 0) {
            List<Integer> sol = new ArrayList();
            for (int num : stack) {
                sol.add(num);
            }
            solutions.add(sol);
            return;
        }
        for (int index : new HashSet<Integer>(unusedIndices)) {
            stack.push(nums[index]);
            unusedIndices.remove(index);
            backtrack();
            unusedIndices.add(index);
            stack.pop();
        }
    }
}

File: python/permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        #keep track of progress outside of the recursion flow
        self.nums = nums
        self.solutions = []
        self.unused_indices = set([i for i in range(len(nums))])
        self.stack = []
        self.backtrack()
        return self.solutions
    
    def backtrack(self) -> None:
        #recursion base case: if stack is complete, add it to solutions
        if len(self.unused_indices) == 0:
            self.solutions.append(self.stack[:])
            return
        #for each unused index, add the next corresponding value of nums to the stack before calling backtrack again
        for i in self.unused_indices.copy():
            self.stack.append(self.nums[i])
            self.unused_indices.remove(i)
            self.backtrack()
            self.unused_indices.add(i)
            self.stack.pop()